### PR TITLE
Nuxt 3 - Quickstart

### PR DESCRIPTION
**Pull request by alex-fusionauth**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207081161912884

### CODE REVIEW

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+Also copy an image file into a new directory within `public` called `img`.
+
+```shell
+mkdir ./public/img && cp ../complete-application/public/img/money.jpg ./public/img/money.jpg
+```
+
+## Authentication
+
+Update `nuxt.config.ts` to include the new `@sidebase/nuxt-auth` and include the provider type of `authjs` just like below. Please note if you would like to code with any of the pages prior to enabling authentication, just set `globalAppMiddleware` to false.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:90]: The "Please note..." sentence should be in an Aside with type = note.

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/nuxt.config.ts"}
+  lang="ts" />
+
+Nuxt 3 includes <a href="https://nuxt.com/docs/guide/directory-structure/server#server-routes" target="_blank">Route Handlers</a>, which are the preferred way to handle REST-like requests. In the `Changebank` application you can configure NextAuth.js FusionAuth's provider in a new route handler by creating a file within `/server/api/auth/[...].ts`.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:95]: As not a Nuxt developer, the file name [...].ts is a little weird. If this is normal in Nuxt then leave, if not, maybe make another note that the file name contains the brackets and dots.

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+Create a new file named `/server/api/auth/[...].ts` and copy the following code for the ChangeBank application.
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/server/api/auth/[...].ts"}
+  lang="ts" />
+
+## App Customization
+
+### Styles
+
+Create a new file named `/complete-application/assets/css/global.css` and copy the below CSS for the ChangeBank application.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:110]: Should this be in the `changebank` directory instead of complete-application?

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/assets/css/global.css"}
+  lang="css" />
+
+The above styles will be imported into a new layout that you will create called `/complete-application/layouts/default.vue`. Update this file to match the code below. This will allow the rest of the pages within our application and the layout itself to use the global styles. This is the first time that you will use the [useAuth](https://sidebase.io/nuxt-auth/v0.6/application-side/session-access-and-management) Composable from Sidebase. The status can then be used to determine what links should be shown in the layout.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:115]: changebank or complete-application?

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/layouts/default.vue"}
+  lang="ts" />
+
+### Login Button
+
+Create a new file in `/complete-application/components/LoginButton.vue` that will be used for a button component.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:122]: changeback or complete-application?

check all pages

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+One special note here is that if you leave the auth `globalAppMiddleware` enabled in `nuxt.config.ts` you will need to authenticate to access this page.
+
+Here's the contents of `/complete-application/pages/account.vue`.
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/pages/account.vue"}
+  lang="ts" />
+
+### Make Change Page
+
+Finally, we'll add some business logic for logged in users to make change with the following code in `/complete-application/pages/account.vue`:
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:158]: Am I just supposed to add this to the bottom of the account .vue page? I think it should be makechange.vue

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+Create a new file in `/complete-application/components/LoginButton.vue` that will be used for a button component.
+
+For this button you can use the `signIn` and `signOut` functions from the `useAuth` composable. Copy the below code for the ChangeBank application into `/complete-application/components/LoginButton.vue`.
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/components/LoginButton.vue"}
+  lang="ts" />
+
+### Login Link
+
+Create a new file in `/complete-application/components/LoginLink.vue` that will be used for a link component. For this link you can use the the same `signIn` function from the `useAuth` composable as you did in the button above.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:131]:
```suggestion
Create a new file in `/complete-application/components/LoginLink.vue` that will be used for a link component. For this link you can use the same `signIn` function from the `useAuth` composable as you did in the button above.
```

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  lang="ts" />
+
+### Make Change Page
+
+Finally, we'll add some business logic for logged in users to make change with the following code in `/complete-application/pages/account.vue`:
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/pages/makechange.vue"}
+  lang="ts" />
+
+If the user session is not present the user is redirect back to the homepage at the base route. If the user is present then the `MakeChangeForm` is presented. Create a new file located at `/complete-application/components/MakeChangeForm.vue` with the below code. This component has all of the business logic needed for taking in a dollar amount of money and returning the correct amount of each coin.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:163]:
```suggestion
If the user session is not present the user is redirected back to the homepage at the base route. If the user is present then the `MakeChangeForm` is presented. Create a new file located at `/complete-application/components/MakeChangeForm.vue` with the below code. This component has all of the business logic needed for taking in a dollar amount of money and returning the correct amount of each coin.
```

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+You can now open up an incognito window and visit the Nuxt app at http://localhost:3000/. Log in with the user account you created when setting up FusionAuth, and you’ll see the email of the user next to a logout button.
+
+```shell
+npm run dev
+```
+
+Try clicking the `Login` button at the top or the link at the center of the screen.
+
+This will take you through the `NextAuth.js` authentication flow. Automatically presenting the FusionAuth Login page because we have `fusionauth` set as the provider in our Login Button `signIn('fusionauth', { callbackUrl: '/account' })`. If you would like to see all providers you can set `fusionauth` to `undefined`.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:178]:
```suggestion
This will take you through the `NextAuth.js` authentication flow. Since `fusionauth` is set as the provider in the Login Button `signIn('fusionauth', { callbackUrl: '/account' })`, the FusionAuth Login page will be automatically presented. If you would like to see all providers you can set `fusionauth` to `undefined`.
```

**Review by mark-robustelli — CHANGES_REQUESTED** [review]
I was not able to get the application to run or the complete application. There is some confusion to me on what should be in changebank and what should be in complete-application.  Also there should be a note or section that would just allow me to npm run dev from the complete-application directory without any changes.  I will try again after the updates.

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+cd changebank
+```
+
+Install [@sidebase/nuxt-auth](https://sidebase.io/nuxt-auth/v0.6/getting-started/installation), which simplifies integrating with FusionAuth and creating a secure web application.
+
+```shell
+npm install next-auth
+```
+
+Copy environment variables from our complete application example.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:76]:
```suggestion
Copy environment variables from the complete application example.
```

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+<Aside type="note">
+If you would like to code with any of the pages prior to enabling authentication, just set `globalAppMiddleware` to false.
+</Aside>
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/nuxt.config.ts"}
+  lang="ts" />
+
+Nuxt 3 includes <a href="https://nuxt.com/docs/guide/directory-structure/server#server-routes" target="_blank">Route Handlers</a>, which are the preferred way to handle REST-like requests. In the `Changebank` application you can configure NextAuth.js FusionAuth's provider in a new route handler by creating a file within `/server/api/auth/[...].ts`.
+
+On first load of Nuxt this file will make sure that you have all of the correct environment variables. The variables are then exported in an object called `authOptions` which can be imported on the server when you need to get our session using `getServerSession`.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:101]:
```suggestion
On first load of Nuxt this file will make sure that you have all of the correct environment variables. The variables are then exported in an object called `authOptions` which can be imported on the server when you need to get your session using `getServerSession`.
```

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+## App Customization
+
+### Styles
+
+Create a new file named `/changebank/assets/css/global.css` and copy the below CSS for the ChangeBank application.
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/assets/css/global.css"}
+  lang="css" />
+
+The above styles will be imported into a new layout that you will create called `/changebank/layouts/default.vue`. Update this file to match the code below. This will allow the rest of the pages within our application and the layout itself to use the global styles. This is the first time that you will use the [useAuth](https://sidebase.io/nuxt-auth/v0.6/application-side/session-access-and-management) Composable from Sidebase. The status can then be used to determine what links should be shown in the layout.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:119]:
```suggestion
The above styles will be imported into a new layout that you will create called `/changebank/layouts/default.vue`. Update this file to match the code below. This will allow the rest of the pages within your application and the layout itself to use the global styles. This is the first time that you will use the [useAuth](https://sidebase.io/nuxt-auth/v0.6/application-side/session-access-and-management) Composable from Sidebase. The status can then be used to determine what links should be shown in the layout.
```

File: astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+Create a new file in `/changebank/components/LoginLink.vue` that will be used for a link component. For this link you can use the same `signIn` function from the `useAuth` composable as you did in the button above.
+
+Copy the below code for the ChangeBank application into `/changebank/components/LoginLink.vue`.
+
+<RemoteCode url={frontmatter.codeRoot + "/complete-application/components/LoginLink.vue"}
+  lang="ts" />
+
+### Home Page
+
+Update the file `/changebank/pages/index.vue` which will have the Homepage details. Not much here just an image and a Login Link.
```
Comment (mark-robustelli) [inline astro/src/content/quickstarts/quickstart-javascript-nuxt-web.mdx:144]: This did not exist, had to create.

**Review by mark-robustelli — CHANGES_REQUESTED** [review]
a few changes.  Include the other changes we spoke about and it should be good.

**Review by mark-robustelli — APPROVED** [review]
Still have some Vale changes that need to be fixed.  Also, I don't think this is a show stopper and I haven't tested with other quickstarts, but if you are logged in and you open another brower and got the application, part of the page recognized that you are logged in and part does not. 

Other than that, should be good.

Screenshot 2024-05-09 at 12 54 28 PM

**Comment by alex-fusionauth** [comment]
Vale changes are for sure bogus. I hate the fact that there are so many false positives. It is catching things in links. I am seeing mostly what I would see as suggestions at best.

I think what you are running into is that the homepage on these can be hit and doesn't force a redirect. I would be happy to update but didn't want to confuse the user too much. We can also use things like middleware on the server side to guarantee a lot more of this as well instead of through the meta layer. 

Basically the flow goes 
- Login -> FusionAuth -> Redirect Success -> Account
Opening a tab / window
- Homepage -> Account
What can't happen is 
- Logout -> Homepage -> Account

**Comment by alex-fusionauth** [comment]
@mark-robustelli can you approve all the requested changes if you are good?
